src/Parallafka/MessageCommitter.cs

```diff
@@ -1,126 +1,93 @@
 ﻿using System;
 using System.Threading;
 using System.Threading.Tasks;
-using System.Threading.Tasks.Dataflow;
 using Microsoft.Extensions.Logging;
 using Parallafka.KafkaConsumer;
 
 namespace Parallafka
 {
-    internal class MessageCommitter<TKey, TValue>
+    /// <summary>
+    /// Commits offsets to Kafka
+    /// </summary>
+    /// <typeparam name="TKey">The key type</typeparam>
+    /// <typeparam name="TValue">The value type</typeparam>
+    internal class MessageCommitter<TKey, TValue> : IMessageCommitter
     {
         private readonly IKafkaConsumer<TKey, TValue> _consumer;
-        private readonly CommitState<TKey, TValue> _commitState;
+        private readonly IMessagesToCommit<TKey, TValue> _commitState;
         private readonly ILogger _logger;
-        private readonly CancellationToken _stopToken;
-        private readonly CancellationTokenSource _stopTimer;
-        private readonly ActionBlock<int> _commitBlock;
+        private readonly SemaphoreSlim _committerLock;
+
+        private long _messagesCommitted;
+        private long _messagesCommitErrors;
+        private long _messagesCommitLoops;
 
         public MessageCommitter(
             IKafkaConsumer<TKey, TValue> consumer,
-            CommitState<TKey, TValue> commitState,
-            ILogger logger,
-            TimeSpan timerDelay,
-            CancellationToken stopToken)
+            IMessagesToCommit<TKey, TValue> commitState,
+            ILogger logger)
         {
-            this._commitBlock = new ActionBlock<int>(_ => GetAndCommitAnyMessages(),
-                new ExecutionDataflowBlockOptions
-                {
-                    MaxDegreeOfParallelism = 1,
-                    BoundedCapacity = 2
-                });
-
+            this._committerLock = new SemaphoreSlim(1, 1);
```

Comment:
Aside: TIL maxCount... is it just me or is it odd that without it `Release()` expands the total available beyond initialCount? I would have expected the single arg constructor to take maxCount

Reply:
Yeah.  I think the single-arg constructor shouldn't exist: code should define the maxCount every time.